### DISCUSSION TITLE
Multicursor and preserving replacement feature

### DISCUSSION THREAD

**Discussion by johmsalas** [johmsalas/text-case.nvim#125] [Ideas]
Inspired by https://github.com/microsoft/vscode/issues/106502
There is a plugin for vscode https://marketplace.visualstudio.com/items?itemName=Cardinal90.multi-cursor-case-preserve

Compared to the existing VsCode solution, if this feature is implemented in text-case.nvim it can use all the supported Text Cases in the plugin. 

**Replacement considerations**

- When [ctrl+d](https://github.com/johmsalas/text-case.nvim/blob/c22d4572c82a14e773a189d50b8182b2a92f8d59/lua/textcase/plugin/multicursor.lua#L169C1-L169C1) (customizable keybinding) is pressed it will [highlight the next occurrence](https://github.com/johmsalas/text-case.nvim/blob/c22d4572c82a14e773a189d50b8182b2a92f8d59/lua/textcase/plugin/multicursor.lua#L39), even if the text case is different as long as the base is the same
- Support for visual mode. It will [use the selected text](https://github.com/johmsalas/text-case.nvim/blob/c22d4572c82a14e773a189d50b8182b2a92f8d59/lua/textcase/plugin/multicursor.lua#L44C15-L44C32) as the base for the considered text variations
- Support for normal mode. It will [compute the current word](https://github.com/johmsalas/text-case.nvim/blob/c22d4572c82a14e773a189d50b8182b2a92f8d59/lua/textcase/plugin/multicursor.lua#L61). Taking into account the current word could have characters like _, -, . or /
- Multiline support adds extra complexity, I don't think there is a common use case for this

**Multicursor considerations**
* It would include custom multicursor selection because regular multicursor plugins/implementations are decoupled from text cases, they do not select different text cases (it becomes unsupported when the text cases are in snake, pascal, ...)
* The multicursor would only support few operations. So far the only known operation is Change Preserving Case. No delete, change, etc... regular keybindings won't work on the selected cursors.
* Once the locations are selected using ctrl+d. There are 2 ways of triggering the Change Preserving Case
  * Pressing c. This is the natural way, but requires to register a keymap on the letter c on visual mode. This could be disruptive to the original c keybinding. Perhaps indications using [Hydra](https://github.com/anuvyklack/hydra.nvim)?
  * A custom keybinding, ie gax. Unnatural developer experience

Created a draft to outline the code design: https://github.com/johmsalas/text-case.nvim/pull/124
Will update this desc when there are more conclusions to keep track of them

**Comment by johmsalas** [comment]
The proof of concept worked, the code has not been organized and there are several things not working yet, ie the text preserving is hard coded to camel case

Some conclusions so far are: 
- Inherit from the following highlight groups: Search and IncSearch
- Currently, the follow keymaps are used:
  - `ctrl+d` select initial instance and next match, also to find new matches
  - `gax` starts preserving case selection
  - `ga<esc>` exits multi-cursor
  - All these keymaps could be changed to easier keymaps like c or <esc> using the same technique as Hydra
- **This is a good moment to evaluate if this feature adds value**.
  -  My first impression is there are already multicursor plugins
  - On the other hand they don't check different cases, that's kind of in the scope of text-case. Would like an opinion on this
  -  Also to consider, Subs replacement is similar to this feature, the main difference is multicursor will bring more grained control. For example, using `ctrl+d` will match next occurrences, and there could be a `ctrl+s` to skip the current cursor jumping to the next match and `ctrl+D` to go backwards

Do you have thoughts about this @peterfication?

![WhatsApp GIF 2023-12-22 at 14 08 46](https://github.com/johmsalas/text-case.nvim/assets/6575405/79fd2518-acc9-442b-b16f-a443aac6c5a5)

**Reply by peterfication** [reply]
Mhh I probably have to think about it more.

My first thought is that I'm not sure if I have a use case for this. Like you said, it's like the `Subs` command somehow. When I really want multiple cursors is when there are a bunch of similar lines I want to make the same changes to that are more than just a replacement. And then the cases are the same (in my experience).
